contracts/src/token/erc1155/mod.rs

```diff
@@ -0,0 +1,1975 @@
+    /// interface id or returned with error, then the error
+    /// [`Error::InvalidReceiver`] is returned.
+    /// If [`IERC1155Receiver::on_erc_1155_batch_received`] hasn't returned its
+    /// interface id or returned with error, then the error
+    /// [`Error::InvalidReceiver`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits a [`TransferSingle`] event if the arrays contain one element, and
+    /// [`TransferBatch`] otherwise.
+    ///
+    /// # Panics
+    ///
+    /// Should not panic.
+    fn do_safe_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        ids: Vec<U256>,
+        values: Vec<U256>,
+        data: &Bytes,
+    ) -> Result<(), Error> {
+        if to.is_zero() {
+            return Err(Error::InvalidReceiver(ERC1155InvalidReceiver {
+                receiver: to,
+            }));
+        }
+        if from.is_zero() {
+            return Err(Error::InvalidSender(ERC1155InvalidSender {
+                sender: from,
+            }));
+        }
+        self._update_with_acceptance_check(from, to, ids, values, data)
+    }
+
+    /// Transfers a `value` amount of `token_id` from `from` to
+    /// `to`. Will mint (or burn) if `from` (or `to`) is the `Address::ZERO`.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `from` - Account to transfer tokens from.
+    /// * `to` - Account of the recipient.
+    /// * `token_id` - Token id.
+    /// * `value` - Amount of tokens to be transferred.
+    ///
+    /// # Errors
+    ///
+    /// If `value` is greater than the balance of the `from` account,
+    /// then the error [`Error::InsufficientBalance`] is returned.
+    ///
+    /// # Panics
+    ///
+    /// If updated balance exceeds `U256::MAX`, may happen during `mint`
+    /// operation.
+    fn do_update(
+        &mut self,
+        from: Address,
+        to: Address,
+        token_id: U256,
+        value: U256,
+    ) -> Result<(), Error> {
+        if !from.is_zero() {
+            let from_balance = self.balance_of(from, token_id);
+            if from_balance < value {
+                return Err(Error::InsufficientBalance(
+                    ERC1155InsufficientBalance {
+                        sender: from,
+                        balance: from_balance,
+                        needed: value,
+                        token_id,
+                    },
+                ));
+            }
+            self._balances
+                .setter(token_id)
+                .setter(from)
+                .sub_assign_unchecked(value);
+        }
+
+        if !to.is_zero() {
+            let new_balance = self
+                ._balances
+                .setter(token_id)
+                .setter(to)
+                .checked_add(value)
+                .expect("should not exceed `U256::MAX` for `_balances`");
+            self._balances.setter(token_id).setter(to).set(new_balance);
```

Comment:
I see we shouldn't call `_balances.setter(..).setter(..)` two times. Just call `getter` method
```suggestion
            let balance = self._balances.getter(token_id).get(to);
            let new_balance = balance
                .checked_add(value)
                .expect("should not exceed `U256::MAX` for `_balances`");
            self._balances.setter(token_id).setter(to).set(new_balance);
```